backend/cfg/cfg_irc_utils.ml

```diff
@@ -242,103 +242,147 @@ module Spilling_heuristics = struct
             (available_heuristics ())))
 end
 
-(* CR xclerc for xclerc: consider dynamic sorted arrays *)
-module WorkList = struct
+module ArraySet = struct
   module type S = sig
     type e
 
     type t
 
-    module Set : Set.S with type elt = e
+    val make : original_capacity:int -> t
 
-    val make : expected_max_size:int -> t
-
-    val empty : t -> t
+    val clear : t -> unit
 
     val is_empty : t -> bool
 
-    val add : t -> e -> t
+    val choose_and_remove : t -> e option
 
-    val remove : t -> e -> t
+    val add : t -> e -> unit
 
-    val choose_and_remove : t -> (e * t) option
+    val remove : t -> e -> unit
 
     val iter : t -> f:(e -> unit) -> unit
 
-    val fold : t -> f:(e -> 'a -> 'a) -> init:'a -> 'a
+    val fold : t -> f:('a -> e -> 'a) -> init:'a -> 'a
 
     val to_list : t -> e list
-
-    val to_set : t -> Set.t
   end
 
-  module Make (E : Set.OrderedType) (ES : Set.S with type elt = E.t) :
-    S with type e = E.t and module Set = ES = struct
-    module Set = ES
-
-    let cut_off = 16
-
-    type e = E.t
-
-    type t =
-      | List of e List.t
-      | Set of Set.t
+  module type OrderedTypeWithDummy = sig
+    include Set.OrderedType
 
-    let empty_list = List []
-
-    let empty_set = Set Set.empty
+    val dummy : t
+  end
 
-    let make ~expected_max_size =
-      if expected_max_size < cut_off then empty_list else empty_set
+  (* CR-someday xclerc for xclerc: consider using unsafe versions of blit and
+     fill. *)
 
-    let empty t = match t with List _ -> empty_list | Set _ -> empty_set
+  module Make (T : OrderedTypeWithDummy) : S with type e = T.t = struct
+    type e = T.t
 
-    let is_empty t =
-      match t with
-      | List l -> ( match l with [] -> true | _ :: _ -> false)
-      | Set s -> Set.is_empty s
+    type t =
+      { mutable array : e array;
+        mutable length : int
+      }
+
+    let make ~original_capacity =
+      let array = Array.make (max 1 original_capacity) T.dummy in
+      let length = 0 in
+      { array; length }
+
+    let clear t =
+      Array.fill t.array ~pos:0 ~len:t.length T.dummy;
+      t.length <- 0
+
+    let is_empty t = Int.equal t.length 0
+
+    let index array length e =
```

Comment:
I think it was originally to avoid field
accesses, but looking at the callers it
no{t, longer} obvious to me CSE will
ensure only one access. And that
would hardly make a difference, anyway.

Reply:
yes, it's a very minor comment, but it would be clearer with `t` and I think it would avoid confusion about whether `length` argument is `Array.length t.array` or `t.length`.